File: src/mcts.js
import { BOARD_SIZE, OASIS_POS } from './constants.js';
import { getValidSlideMoves, getValidLShapeMoves, checkWinCondition } from './engine.js';
import { evaluateBoardV7 } from './heuristics.v7.js';

class MCTSNode {
    constructor(state, player, priorProb, parent = null, move = null) {
        this.state = state; // 2D array
        this.player = player;
        this.priorProb = priorProb;
        this.parent = parent;
        this.move = move;
        
        this.children = new Map(); // key: string "px,py->tx,ty", value: MCTSNode
        this.visitCount = 0;
        this.valueSum = 0.0;
    }

    get qValue() {
        if (this.visitCount === 0) return 0;
        return this.valueSum / this.visitCount;
    }

    expand(actionProbs) {
        // actionProbs: Array of { move, prob }
        for (const { move, prob } of actionProbs) {
            const nextState = this.simulateMove(this.state, move);
            const nextPlayer = this.player === 1 ? 2 : 1;
            const moveKey = `${move.pieceX},${move.pieceY}->${move.targetX},${move.targetY}`;
            
            if (!this.children.has(moveKey)) {
                this.children.set(moveKey, new MCTSNode(
                    nextState,
                    nextPlayer,
                    prob,
                    this,
                    move
                ));
            }
        }
    }

    simulateMove(board, move) {
        const nextBoard = board.map(row => [...row]);
        const p = nextBoard[move.pieceY][move.pieceX];
        nextBoard[move.pieceY][move.pieceX] = 0;
        nextBoard[move.targetY][move.targetX] = p;
        return nextBoard;
    }

    isExpanded() {
        return this.children.size > 0;
    }
}

export class MCTS {
    constructor(onnxSession, ort, cPuct = 2.5, numSimulations = 100) {
        this.session = onnxSession;
        this.ort = ort;
        this.cPuct = cPuct;
        this.numSimulations = numSimulations;
        this.nnCache = new Map(); // key: JSON.stringify(board)+player
    }

    async getActionProbs(board, player) {
        const stateKey = JSON.stringify(board) + player;
        if (this.nnCache.has(stateKey)) return this.nnCache.get(stateKey);

        const possibleMoves = this.getAllPossibleMoves(board, player);
        if (possibleMoves.length === 0) return [];

        // Prepare ONNX input (1, 3, 11, 11)
        const inputData = this.prepareInput(board, player);
        const inputTensor = new this.ort.Tensor('float32', inputData, [1, 3, 11, 11]);
        
        const results = await this.session.run({ input: inputTensor });
        const logitsFrom = results.logits_from.data; // Float32Array(121)
        const logitsTo = results.logits_to.data;     // Float32Array(121)

        // Masking logic
        const fromMask = new Uint8Array(121).fill(0);
        for (const m of possibleMoves) {
            fromMask[m.pieceY * BOARD_SIZE + m.pieceX] = 1;
        }

        // Softmax for From
        const probsFrom = this.softmaxWithMask(logitsFrom, fromMask);
        // Softmax for To (simplified, since we multiply later)
        const probsTo = this.softmax(logitsTo);

        let actionProbs = [];
        for (const move of possibleMoves) {
            const fromIdx = move.pieceY * BOARD_SIZE + move.pieceX;
            const toIdx = move.targetY * BOARD_SIZE + move.targetX;
            const prob = probsFrom[fromIdx] * probsTo[toIdx];
            actionProbs.push({ move, prob });
        }

        // Normalize
        const totalProb = actionProbs.reduce((sum, item) => sum + item.prob, 0);
        if (totalProb > 0) {
            actionProbs = actionProbs.map(item => ({ ...item, prob: item.prob / totalProb }));
        }

        this.nnCache.set(stateKey, actionProbs);
        return actionProbs;
    }

    prepareInput(board, currentPlayer) {
        const opponent = currentPlayer === 1 ? 2 : 1;
        const data = new Float32Array(3 * 11 * 11);
        
        const MEADOW_SET = new Set([
            "5,4", "5,6", "4,5", "6,5",
            "5,3", "5,7", "3,5", "7,5",
            "4,4", "6,4", "4,6", "6,6"
        ]);

        for (let y = 0; y < 11; y++) {
            for (let x = 0; x < 11; x++) {
                const val = board[y][x];
                // Channel 0: Current player
                if (val === currentPlayer) data[0 * 121 + y * 11 + x] = 1.0;
                // Channel 1: Opponent
                else if (val === opponent) data[1 * 121 + y * 11 + x] = 1.0;
                
                // Channel 2: Special tiles
                if (x === 5 && y === 5) data[2 * 121 + y * 11 + x] = 1.0;
                else if (MEADOW_SET.has(`${x},${y}`)) data[2 * 121 + y * 11 + x] = 0.5;
            }
        }
        return data;
    }

    softmaxWithMask(logits, mask) {
        const result = new Float32Array(logits.length);
        let maxVal = -Infinity;
        for (let i = 0; i < logits.length; i++) {
            if (mask[i] && logits[i] > maxVal) maxVal = logits[i];
        }
        
        let sumExp = 0;
        for (let i = 0; i < logits.length; i++) {
            if (mask[i]) {
                result[i] = Math.exp(logits[i] - maxVal);
                sumExp += result[i];
            } else {
                result[i] = 0;
            }
        }
        for (let i = 0; i < logits.length; i++) {
            result[i] /= sumExp;
        }
        return result;
    }

    softmax(logits) {
        const result = new Float32Array(logits.length);
        let maxVal = -Infinity;
        for (let i = 0; i < logits.length; i++) {
            if (logits[i] > maxVal) maxVal = logits[i];
        }
        let sumExp = 0;
        for (let i = 0; i < logits.length; i++) {
            result[i] = Math.exp(logits[i] - maxVal);
            sumExp += result[i];
        }
        for (let i = 0; i < logits.length; i++) {
            result[i] /= sumExp;
        }
        return result;
    }

    getAllPossibleMoves(board, player) {
        const moves = [];
        for (let y = 0; y < BOARD_SIZE; y++) {
            for (let x = 0; x < BOARD_SIZE; x++) {
                if (board[y][x] === player) {
                    const slide = getValidSlideMoves(board, x, y);
                    const lshape = getValidLShapeMoves(board, x, y);
                    for (const m of [...slide, ...lshape]) {
                        moves.push({
                            pieceX: x, pieceY: y,
                            targetX: m.x, targetY: m.y,
                            type: m.type, player: player
                        });
                    }
                }
            }
        }
        return moves;
    }

    evaluateNode(board, evalPlayer) {
        const opponent = evalPlayer === 1 ? 2 : 1;
        const rawScore = evaluateBoardV7(board, evalPlayer, opponent);
        return Math.tanh(rawScore / 2000.0);
    }

    async search(initialBoard, player) {
        const root = new MCTSNode(initialBoard, player, 1.0);
        
        // Initial expansion
        const actionProbs = await this.getActionProbs(initialBoard, player);
        root.expand(actionProbs);

        if (!root.isExpanded()) return null;

        for (let sim = 0; sim < this.numSimulations; sim++) {
            let node = root;
            const searchPath = [node];

            // 1. Selection
            while (node.isExpanded()) {
                let bestScore = -Infinity;
                let bestChild = null;
                
                for (const child of node.children.values()) {
                    const q = -child.qValue;
                    const u = this.cPuct * child.priorProb * Math.sqrt(node.visitCount) / (1 + child.visitCount);
                    const score = q + u;
                    if (score > bestScore) {
                        bestScore = score;
                        bestChild = child;
                    }
                }
                node = bestChild;
                searchPath.push(node);
            }

            // 2. Expansion / Evaluation
            let isTerminal = false;
            let value = 0.0;
            
            if (node.move && checkWinCondition(node.move.targetX, node.move.targetY)) {
                isTerminal = true;
                value = -1.0;
            }
            
            if (!isTerminal) {
                const probs = await this.getActionProbs(node.state, node.player);
                if (probs.length === 0) {
                    value = -1.0;
                } else {
                    node.expand(probs);
                    value = this.evaluateNode(node.state, node.player);
                }
            }

            // 3. Backpropagation
            for (let i = searchPath.length - 1; i >= 0; i--) {
                const n = searchPath[i];
                n.valueSum += value;
                n.visitCount += 1;
                value = -value;
            }
        }

        // Return best move by visit count
        let bestChild = null;
        let maxVisits = -1;
        for (const child of root.children.values()) {
            if (child.visitCount > maxVisits) {
                maxVisits = child.visitCount;
                bestChild = child;
            }
        }
        return bestChild ? bestChild.move : null;
    }
}

File: backend/mcts.py
"""
Monte Carlo Tree Search (MCTS) implementation for AI Policy Network V4.
Integrates PUCT selection, NN-based expansion w/ caching, and heuristic evaluation w/ tanh normalization.
"""
import math
import torch
import numpy as np

from engine import BOARD_SIZE, check_win_condition
from ai import get_all_possible_moves, simulate_move, evaluate_board

try:
    from model_v4 import board_to_channels, get_v4_masks
except ImportError:
    pass

def hash_board(board, player):
    """Create a hashable representation of the board state for caching."""
    return tuple(tuple(row) for row in board), player

class MCTSNode:
    def __init__(self, state, player, prior_prob, parent=None, move=None):
        self.state = state
        self.player = player
        self.prior_prob = prior_prob
        self.parent = parent
        self.move = move
        
        self.children = {}
        self.visit_count = 0
        self.value_sum = 0.0

    @property
    def q_value(self):
        if self.visit_count == 0:
            return 0
        return self.value_sum / self.visit_count

    def expand(self, action_probs):
        """action_probs is a list of tuples: (move_dict, probability)"""
        for move, prob in action_probs:
            next_state = simulate_move(self.state, move)
            next_player = 2 if self.player == 1 else 1
            
            # Unique key for the move based on coordinates
            move_key = (move["pieceX"], move["pieceY"], move["targetX"], move["targetY"])
            
            if move_key not in self.children:
                self.children[move_key] = MCTSNode(
                    state=next_state,
                    player=next_player,
                    prior_prob=prob,
                    parent=self,
                    move=move
                )

    def is_expanded(self):
        return len(self.children) > 0

class MCTS:
    def __init__(self, model, device, c_puct=2.5, num_simulations=100):
        self.model = model
        self.device = device
        self.c_puct = c_puct
        self.num_simulations = num_simulations
        self.nn_cache = {}  # Inference speedup: caches (board, player) -> action_probs

    def get_action_probs(self, board, player):
        state_key = hash_board(board, player)
        if state_key in self.nn_cache:
            return self.nn_cache[state_key]

        possible_moves = get_all_possible_moves(board, player)
        if not possible_moves:
            return []

        channels = board_to_channels(board, player).unsqueeze(0).to(self.device)
        from_mask, _ = get_v4_masks(board, player)
        from_mask = from_mask.to(self.device)

        with torch.no_grad():
            logits_from, logits_to = self.model(channels)
            logits_from[0][~from_mask] = -1e9
            
            probs_from = torch.nn.functional.softmax(logits_from[0], dim=0)
            probs_to = torch.nn.functional.softmax(logits_to[0], dim=0)

        action_probs = []
        for move in possible_moves:
            from_idx = move["pieceY"] * BOARD_SIZE + move["pieceX"]
            to_idx = move["targetY"] * BOARD_SIZE + move["targetX"]
            prob = probs_from[from_idx].item() * probs_to[to_idx].item()
            action_probs.append((move, prob))
            
        # Normalize sum to 1.0
        total_prob = sum(p for _, p in action_probs)
        if total_prob > 0:
            action_probs = [(m, p / total_prob) for m, p in action_probs]

        self.nn_cache[state_key] = action_probs
        return action_probs

    def evaluate(self, board, eval_player):
        """
        Uses existing AI heuristic with tanh normalization to [-1, 1].
        """
        opponent = 2 if eval_player == 1 else 1
        raw_score = evaluate_board(board, ai_player=eval_player, human_player=opponent)
        
        # Scaling factor to squash huge heuristic values (WIN=100000, MEADOW=500) into reasonable margins
        return math.tanh(raw_score / 2000.0)

    def search(self, initial_board, player, add_noise=False):
        root = MCTSNode(state=initial_board, player=player, prior_prob=1.0)
        
        # Expand root
        action_probs = self.get_action_probs(initial_board, player)
        root.expand(action_probs)

        if not root.is_expanded():
            return None

        if add_noise and len(root.children) > 0:
            # Dirichlet noise on root to encourage exploration (AlphaGo technique)
            dirichlet_alpha = 0.3
            noise = np.random.dirichlet([dirichlet_alpha] * len(root.children))
            for i, child in enumerate(root.children.values()):
                child.prior_prob = 0.75 * child.prior_prob + 0.25 * noise[i]

        for sim in range(self.num_simulations):
            node = root
            search_path = [node]

            # 1. Selection
            while node.is_expanded():
                best_score = -float('inf')
                best_child = None
                
                for child in node.children.values():
                    # PUCT Formula: Q + c * P * sqrt(N_parent) / (1 + N_child)
                    # We negate child.q_value because Q is from the child's perspective (opponent)
                    q = -child.q_value 
                    u = self.c_puct * child.prior_prob * math.sqrt(node.visit_count) / (1 + child.visit_count)
                    score = q + u

                    if score > best_score:
                        best_score = score
                        best_child = child
                        
                node = best_child
                search_path.append(node)

            # 2. Evaluation / Expansion
            is_terminal = False
            value = 0.0
            
            # Did the last move win the game?
            if node.move and check_win_condition(node.move["targetX"], node.move["targetY"]):
                is_terminal = True
                # The parent player made the winning move. So for the current node's player, they lost (-1.0).
                value = -1.0 
            
            if not is_terminal:
                action_probs = self.get_action_probs(node.state, node.player)
                if not action_probs:
                    value = -1.0 # Terminal loss due to no moves
                else:
                    node.expand(action_probs)
                    value = self.evaluate(node.state, node.player)
            
            # 3. Backpropagation
            # The value flips at each step because it's a zero-sum turn-based game
            for n in reversed(search_path):
                n.value_sum += value
                n.visit_count += 1
                value = -value

        # Return the move that was visited the most (most robust search branch)
        best_child = max(root.children.values(), key=lambda c: c.visit_count)
        return best_child.move
